AmplifyPlugins/API/Sources/AWSAPIPlugin/Operation/AWSGraphQLSubscriptionTaskRunner.swift

```diff
@@ -378,6 +378,31 @@ fileprivate func toAPIError<R: Decodable>(_ errors: [Error], type: R.Type) -> AP
         (hasAuthorizationError ? ": \(APIError.UnauthorizedMessageString)" : "")
     }
 
+#if swift(<5.8)
+    if let errors = errors.cast(to: AppSyncRealTimeRequest.Error.self) {
+        let hasAuthorizationError = errors.contains(where: { $0 == .unauthorized})
+        return APIError.operationError(
+            errorDescription(hasAuthorizationError),
+            "",
+            errors.first
+        )
+    } else if let errors = errors.cast(to: GraphQLError.self) {
+        let hasAuthorizationError = errors.map(\.extensions)
+            .compactMap { $0.flatMap { $0["errorType"]?.stringValue } }
+            .contains(where: { AppSyncErrorType($0) == .unauthorized })
+        return APIError.operationError(
+            errorDescription(hasAuthorizationError),
+            "",
+            GraphQLResponseError<R>.error(errors)
+        )
+    } else {
+        return APIError.operationError(
+            errorDescription(),
+            "",
+            errors.first
+        )
+    }
+#else
```

Comment:
looks like you can probably reduce this once more
```swift
    let hasAuthorizationError: Bool
    let underlyingError: Error?
#if swift(<5.8)
    if let errors = errors.cast(to: AppSyncRealTimeRequest.Error.self) {
        hasAuthorizationError = errors.contains(where: { $0 == .unauthorized})
        underlyingError = errors.first
    } else if let errors = errors.cast(to: GraphQLError.self) {
        hasAuthorizationError = errors.map(\.extensions)
            .compactMap { $0.flatMap { $0["errorType"]?.stringValue } }
            .contains(where: { AppSyncErrorType($0) == .unauthorized })
        underlyingError = GraphQLResponseError<R>.error(errors)
    } else {
        hasAuthorizationError = false
        underlyingError = errors.first
    }
#else
    if let errors = errors as? [AppSyncRealTimeRequest.Error] {
        hasAuthorizationError = errors.contains(where: { $0 == .unauthorized})
        underlyingError = errors.first
    } else if let errors = errors as? [GraphQLError] {
        hasAuthorizationError = errors.map(\.extensions)
            .compactMap { $0.flatMap { $0["errorType"]?.stringValue } }
            .contains(where: { AppSyncErrorType($0) == .unauthorized })
        underlyingError = errors.first
    } else {
        hasAuthorizationError = false
        underlyingError = errors.first
    }
#endif
    return APIError.operationError(
        errorDescription(hasAuthorizationError),
        "",
        underlyingError
    )
```

Or even further, just use `cast` everywhere. This would avoid having to duplicate business logic while maintaining the Collection type casting in one place. Once we support 5.8 and up, then we can replace usages of `cast` with `as`
```
func cast<T>(to type: T.Type) -> [T]? {
        #if swift(<5.8)
        self.reduce([]) { partialResult, ele in
            if let partialResult, let ele = ele as? T {
                return partialResult + [ele]
            }
            return nil
        }
        #else
        return self as? [T]
        #endif
    }
```

Reply:
I like the idea of having the compiler directive in the `cast` function instead

Reply:
Including the directive in the `cast` func will make unit testing harder. Once we upgrade the Swift tool version to 5.8, removing the code within the annotation will become straightforward.

Reply:
Why would it make unit testing harder? We don't need to unit-test that we support both 5.7 and 5.8

Reply:
Chatted with @5d, the CI set up will cause `cast` to be unit tested against Swift 5.9 so it won't reach the code path. We'd have to refactor the unit test to test against something like:
```swift
func cast<T>(to type: T.Type) -> [T]? {
        #if swift(<5.8)
        return castSwift5_7() // update unit tests to test this function instead
        #else
        return self as? [T]
        #endif
    }
```

We to upgrade to min 5.8 soon so as Di mentioned it will be easier to remove the code later in the current PR